File: src/channels/markdown.py
"""Markdown to Telegram HTML converter.

Converts standard Markdown to the HTML subset supported by Telegram Bot API.
Designed for both complete and partial (streaming) content.

Supported Telegram HTML tags:
  <b>, <i>, <s>, <code>, <pre>, <a>, <blockquote>
"""

from __future__ import annotations

import re
from html import escape as html_escape


def md_to_telegram_html(text: str, *, partial: bool = False) -> str:
    """Convert Markdown text to Telegram-compatible HTML.

    Args:
        text: Standard Markdown text.
        partial: If True, force-close unclosed formatting tags (for
            streaming draft updates). On conversion failure, falls back
            to HTML-escaped plain text.

    Returns:
        HTML string safe for Telegram ``parse_mode="HTML"``.
    """
    if not text:
        return ""

    try:
        result = _convert(text, partial=partial)
    except Exception:
        # Fallback: return HTML-escaped plain text
        result = html_escape(text)

    return result


# ---------------------------------------------------------------------------
# Internal conversion
# ---------------------------------------------------------------------------

# Code fence pattern:  ```lang\n...\n```
_CODE_FENCE_OPEN = re.compile(r"^```(\w*)\s*$")
_CODE_FENCE_CLOSE = re.compile(r"^```\s*$")

# Blockquote prefix
_BLOCKQUOTE_PREFIX = re.compile(r"^>\s?(.*)")

# Header pattern
_HEADER = re.compile(r"^(#{1,6})\s+(.*)")

# Table separator row (e.g. |---|---|)
_TABLE_SEP = re.compile(r"^\|[\s:]*-{2,}[\s:]*(\|[\s:]*-{2,}[\s:]*)*\|?\s*$")

# Table row (starts and optionally ends with |)
_TABLE_ROW = re.compile(r"^\|(.+)")


def _convert(text: str, *, partial: bool) -> str:
    lines = text.split("\n")
    output: list[str] = []

    in_code_block = False
    code_lang = ""
    code_lines: list[str] = []
    blockquote_lines: list[str] = []
    table_rows: list[str] = []

    for line in lines:
        # --- Code fence handling ---
        if not in_code_block:
            m = _CODE_FENCE_OPEN.match(line)
            if m:
                # Flush pending blocks
                _flush_blockquote(blockquote_lines, output)
                _flush_table(table_rows, output)
                in_code_block = True
                code_lang = m.group(1)
                code_lines = []
                continue
        else:
            if _CODE_FENCE_CLOSE.match(line):
                _flush_code_block(code_lang, code_lines, output)
                in_code_block = False
                code_lang = ""
                code_lines = []
                continue
            code_lines.append(line)
            continue

        # --- Table handling ---
        if _TABLE_ROW.match(line):
            # Flush blockquote before starting table
            _flush_blockquote(blockquote_lines, output)
            table_rows.append(line)
            continue

        # Flush table if we exit it
        _flush_table(table_rows, output)

        # --- Blockquote handling ---
        bq = _BLOCKQUOTE_PREFIX.match(line)
        if bq:
            blockquote_lines.append(bq.group(1))
            continue

        # Flush blockquote if we exit it
        _flush_blockquote(blockquote_lines, output)

        # --- Header handling ---
        hm = _HEADER.match(line)
        if hm:
            content = _convert_inline(html_escape(hm.group(2)))
            output.append(f"<b>{content}</b>")
            continue

        # --- List item handling (- or *) ---
        list_match = re.match(r"^[-*]\s+(.*)", line)
        if list_match:
            content = _convert_inline(html_escape(list_match.group(1)))
            output.append(f"\u2022 {content}")
            continue

        # --- Regular line: inline formatting ---
        output.append(_convert_inline(html_escape(line)))

    # Flush remaining blocks
    _flush_blockquote(blockquote_lines, output)
    _flush_table(table_rows, output)

    if in_code_block:
        if partial:
            # Force-close unclosed code block
            _flush_code_block(code_lang, code_lines, output)
        else:
            # Treat as regular text
            prefix = f"```{code_lang}" if code_lang else "```"
            output.append(html_escape(prefix))
            for cl in code_lines:
                output.append(html_escape(cl))

    result = "\n".join(output)

    if partial:
        result = _force_close_inline(result)

    return result


def _flush_code_block(
    lang: str,
    lines: list[str],
    output: list[str],
) -> None:
    """Render a code block as <pre><code>...</code></pre>."""
    escaped = html_escape("\n".join(lines))
    if lang:
        output.append(f'<pre><code class="language-{html_escape(lang)}">{escaped}</code></pre>')
    else:
        output.append(f"<pre><code>{escaped}</code></pre>")


def _flush_blockquote(lines: list[str], output: list[str]) -> None:
    """Render accumulated blockquote lines."""
    if not lines:
        return
    content = "\n".join(_convert_inline(html_escape(quote_line)) for quote_line in lines)
    output.append(f"<blockquote>{content}</blockquote>")
    lines.clear()


def _flush_table(rows: list[str], output: list[str]) -> None:
    """Render accumulated table rows as a ``<pre>`` block.

    Telegram HTML does not support ``<table>``, so we render tables as
    monospaced pre-formatted text with aligned columns.
    Markdown formatting markers (``**``, ``~~``) are stripped since they
    cannot render inside ``<pre>`` and would break column alignment.
    """
    if not rows:
        return

    # Parse cells from each row, strip markdown markers for <pre> display
    parsed: list[list[str]] = []
    for row in rows:
        # Skip separator rows (|---|---|)
        if _TABLE_SEP.match(row):
            continue
        cells = [_strip_md_markers(c.strip()) for c in row.strip().strip("|").split("|")]
        parsed.append(cells)

    if not parsed:
        rows.clear()
        return

    # Compute column widths (use display width for CJK support)
    n_cols = max(len(r) for r in parsed)
    col_widths = [0] * n_cols
    for row_cells in parsed:
        for i, cell in enumerate(row_cells):
            if i < n_cols:
                col_widths[i] = max(col_widths[i], _display_width(cell))

    # Render aligned rows
    rendered: list[str] = []
    for idx, row_cells in enumerate(parsed):
        parts = []
        for i in range(n_cols):
            cell = row_cells[i] if i < len(row_cells) else ""
            pad = col_widths[i] - _display_width(cell)
            parts.append(cell + " " * max(0, pad))
        rendered.append("  ".join(parts))

        # Add separator after header row
        if idx == 0 and len(parsed) > 1:
            sep_parts = ["-" * w for w in col_widths]
            rendered.append("  ".join(sep_parts))

    escaped = html_escape("\n".join(rendered))
    output.append(f"<pre>{escaped}</pre>")
    rows.clear()


# ---------------------------------------------------------------------------
# Inline formatting
# ---------------------------------------------------------------------------

# Order matters: process code first (to protect its content), then others.
# Inline code: `code`
_INLINE_CODE = re.compile(r"`([^`]+)`")
# Bold: **text**
_BOLD = re.compile(r"\*\*(.+?)\*\*")
# Italic: *text*  (but not **)
_ITALIC_STAR = re.compile(r"(?<!\*)\*([^*]+?)\*(?!\*)")
# Italic: _text_  (word boundary aware)
_ITALIC_UNDER = re.compile(r"(?<!\w)_([^_]+?)_(?!\w)")
# Strikethrough: ~~text~~
_STRIKE = re.compile(r"~~(.+?)~~")
# Link: [text](url)
_LINK = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")

# Placeholder to protect inline code from further processing
_CODE_PLACEHOLDER = "\x00CODE{}\x00"


def _convert_inline(text: str) -> str:
    """Apply inline Markdown formatting to a single line.

    Input text must already be HTML-escaped.
    """
    # Extract inline code spans first to protect them
    code_spans: list[str] = []

    def _save_code(m: re.Match) -> str:
        idx = len(code_spans)
        code_spans.append(f"<code>{m.group(1)}</code>")
        return _CODE_PLACEHOLDER.format(idx)

    text = _INLINE_CODE.sub(_save_code, text)

    # Apply formatting (on non-code text)
    text = _BOLD.sub(r"<b>\1</b>", text)
    text = _STRIKE.sub(r"<s>\1</s>", text)
    text = _ITALIC_STAR.sub(r"<i>\1</i>", text)
    text = _ITALIC_UNDER.sub(r"<i>\1</i>", text)
    text = _LINK.sub(r'<a href="\2">\1</a>', text)

    # Restore code spans
    for i, span in enumerate(code_spans):
        text = text.replace(_CODE_PLACEHOLDER.format(i), span)

    return text


# ---------------------------------------------------------------------------
# Streaming: force-close unclosed tags
# ---------------------------------------------------------------------------

# Tags we might need to close if they are opened but not closed
_OPEN_CLOSE_TAGS = ["b", "i", "s", "code", "pre", "a", "blockquote"]


def _force_close_inline(html: str) -> str:
    """Ensure all opened HTML tags are closed (for partial streaming).

    Simple approach: count open vs close tags and append missing closes.
    """
    for tag in _OPEN_CLOSE_TAGS:
        open_count = len(re.findall(rf"<{tag}(?:\s[^>]*)?>", html))
        close_count = len(re.findall(rf"</{tag}>", html))
        for _ in range(open_count - close_count):
            html += f"</{tag}>"

    return html


# ---------------------------------------------------------------------------
# Table helpers
# ---------------------------------------------------------------------------

# Markdown formatting markers to strip in <pre> table cells
_MD_MARKERS = re.compile(r"\*\*|~~|__")


def _strip_md_markers(text: str) -> str:
    """Remove Markdown formatting markers (**, ~~, __) from text."""
    return _MD_MARKERS.sub("", text)


def _display_width(text: str) -> int:
    """Approximate display width accounting for CJK double-width chars."""
    width = 0
    for ch in text:
        cp = ord(ch)
        # CJK Unified Ideographs, CJK Compat, Fullwidth forms, etc.
        if (
            0x1100 <= cp <= 0x115F  # Hangul Jamo
            or 0x2E80 <= cp <= 0x9FFF  # CJK radicals through Unified Ideographs
            or 0xAC00 <= cp <= 0xD7AF  # Hangul syllables
            or 0xF900 <= cp <= 0xFAFF  # CJK Compat Ideographs
            or 0xFE10 <= cp <= 0xFE6F  # CJK Compat Forms + Small Forms
            or 0xFF01 <= cp <= 0xFF60  # Fullwidth ASCII
            or 0xFFE0 <= cp <= 0xFFE6  # Fullwidth signs
            or 0x20000 <= cp <= 0x2FA1F  # CJK Extension B-F
        ):
            width += 2
        else:
            width += 1
    return width

File: frontend/src/lib/markdown.ts
/**
 * Minimal Markdown-to-HTML renderer for chat messages.
 *
 * Supports: code fences, tables, headers, bold, italic,
 * strikethrough, inline code, links, blockquotes, lists.
 * No external dependencies.
 */

const ESC: Record<string, string> = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
};

function esc(s: string): string {
  return s.replace(/[&<>"]/g, (ch) => ESC[ch] ?? ch);
}

// ---------- block-level ----------

function renderTable(rows: string[]): string {
  const parsed: string[][] = [];
  for (const row of rows) {
    if (/^\|[\s:]*-{2,}/.test(row)) continue; // skip separator
    const cells = row
      .trim()
      .replace(/^\|/, "")
      .replace(/\|$/, "")
      .split("|")
      .map((c) => c.trim());
    parsed.push(cells);
  }
  if (parsed.length === 0) return "";

  const [header, ...body] = parsed;
  const ths = header.map((c) => `<th>${inlineFormat(esc(c))}</th>`).join("");
  const bodyRows = body
    .map((r) => {
      const tds = r.map((c) => `<td>${inlineFormat(esc(c))}</td>`).join("");
      return `<tr>${tds}</tr>`;
    })
    .join("\n");
  return `<table class="md-table"><thead><tr>${ths}</tr></thead><tbody>${bodyRows}</tbody></table>`;
}

function renderCodeBlock(lang: string, lines: string[]): string {
  const code = esc(lines.join("\n"));
  if (lang) {
    return `<pre><code class="language-${esc(lang)}">${code}</code></pre>`;
  }
  return `<pre><code>${code}</code></pre>`;
}

// ---------- inline ----------

function inlineFormat(text: string): string {
  // inline code (protect first)
  const codeSpans: string[] = [];
  text = text.replace(/`([^`]+)`/g, (_, code) => {
    const idx = codeSpans.length;
    codeSpans.push(`<code>${code}</code>`);
    return `\x00C${idx}\x00`;
  });

  // bold
  text = text.replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>");
  // strikethrough
  text = text.replace(/~~(.+?)~~/g, "<s>$1</s>");
  // italic *text* (not **)
  text = text.replace(/(?<!\*)\*([^*]+?)\*(?!\*)/g, "<em>$1</em>");
  // italic _text_
  text = text.replace(/(?<!\w)_([^_]+?)_(?!\w)/g, "<em>$1</em>");
  // links
  text = text.replace(
    /\[([^\]]+)\]\(([^)]+)\)/g,
    '<a href="$2" target="_blank" rel="noopener">$1</a>',
  );

  // restore code spans
  for (let i = 0; i < codeSpans.length; i++) {
    text = text.replace(`\x00C${i}\x00`, codeSpans[i]);
  }
  return text;
}

// ---------- list grouping ----------

type Block = { type: "html"; html: string } | { type: "li"; tag: "ul" | "ol"; html: string };

function groupLists(blocks: Block[]): string[] {
  const out: string[] = [];
  let currentTag: "ul" | "ol" | null = null;

  const flushList = () => {
    if (currentTag) {
      out.push(`</${currentTag}>`);
      currentTag = null;
    }
  };

  for (const block of blocks) {
    if (block.type === "li") {
      if (currentTag !== block.tag) {
        flushList();
        currentTag = block.tag;
        out.push(`<${currentTag}>`);
      }
      out.push(block.html);
    } else {
      flushList();
      out.push(block.html);
    }
  }
  flushList();
  return out;
}

// ---------- main ----------

export function renderMarkdown(md: string): string {
  const lines = md.split("\n");
  const blocks: Block[] = [];

  let inCode = false;
  let codeLang = "";
  let codeLines: string[] = [];
  let tableRows: string[] = [];
  let bqLines: string[] = [];

  const flushTable = () => {
    if (tableRows.length > 0) {
      blocks.push({ type: "html", html: renderTable(tableRows) });
      tableRows = [];
    }
  };

  const flushBlockquote = () => {
    if (bqLines.length > 0) {
      const inner = bqLines.map((l) => inlineFormat(esc(l))).join("<br>");
      blocks.push({ type: "html", html: `<blockquote>${inner}</blockquote>` });
      bqLines = [];
    }
  };

  for (const line of lines) {
    // Code fence
    if (!inCode && /^```(\w*)/.test(line)) {
      flushTable();
      flushBlockquote();
      inCode = true;
      codeLang = line.replace(/^```/, "").trim();
      codeLines = [];
      continue;
    }
    if (inCode) {
      if (/^```\s*$/.test(line)) {
        blocks.push({ type: "html", html: renderCodeBlock(codeLang, codeLines) });
        inCode = false;
        codeLang = "";
        codeLines = [];
      } else {
        codeLines.push(line);
      }
      continue;
    }

    // Table row
    if (/^\|.+/.test(line)) {
      flushBlockquote();
      tableRows.push(line);
      continue;
    }
    flushTable();

    // Blockquote
    const bqMatch = line.match(/^>\s?(.*)/);
    if (bqMatch) {
      bqLines.push(bqMatch[1]);
      continue;
    }
    flushBlockquote();

    // Header
    const hMatch = line.match(/^(#{1,6})\s+(.*)/);
    if (hMatch) {
      const level = hMatch[1].length;
      blocks.push({ type: "html", html: `<h${level}>${inlineFormat(esc(hMatch[2]))}</h${level}>` });
      continue;
    }

    // Horizontal rule
    if (/^(-{3,}|\*{3,}|_{3,})\s*$/.test(line)) {
      blocks.push({ type: "html", html: "<hr>" });
      continue;
    }

    // Unordered list item
    if (/^[-*]\s+/.test(line)) {
      const content = line.replace(/^[-*]\s+/, "");
      blocks.push({ type: "li", tag: "ul", html: `<li>${inlineFormat(esc(content))}</li>` });
      continue;
    }

    // Ordered list item
    if (/^\d+\.\s+/.test(line)) {
      const content = line.replace(/^\d+\.\s+/, "");
      blocks.push({ type: "li", tag: "ol", html: `<li>${inlineFormat(esc(content))}</li>` });
      continue;
    }

    // Empty line
    if (line.trim() === "") {
      const lastBlock = blocks[blocks.length - 1];
      const lastHtml = lastBlock?.type === "html" ? lastBlock.html : "";
      if (blocks.length > 0 && lastHtml !== '<div class="md-gap"></div>') {
        blocks.push({ type: "html", html: '<div class="md-gap"></div>' });
      }
      continue;
    }

    // Regular text
    blocks.push({ type: "html", html: `<div>${inlineFormat(esc(line))}</div>` });
  }

  // Flush remaining
  flushTable();
  flushBlockquote();
  if (inCode) {
    blocks.push({ type: "html", html: renderCodeBlock(codeLang, codeLines) });
  }

  return groupLists(blocks).join("\n");
}
